File: Miscellaneous/DataStructures/Lab2/tree.py
import numpy as np
'''
Answers to the questions:
Algorithm: See below
3 Test Cases: Any test will function
Running time: Worst case: O(n) Average Case: O(sqrt(n)) Best Case: O(1)
'''
class Node:
	def __init__(self, value, parent=None):
		self.parent = parent
		self.children = [self]
		self.data = value

	def add_children(self, child_node):
		self.children.append(child_node)

class Tree:
	def __init__(self):
		self.root = None
		self.nodes = []
		self.data = []

	def addNode(self, value, parent=None):
		if self.root is None:
			self.root = Node(value)
			self.nodes.append(self.root)
			self.data.append(value)
			print('root added:', value)
		else:
			node = self._find_parent(parent)
			if node != None:
				child = Node(value, parent=node)
				node.add_children(child)
				self.nodes.append(child)
				self.data.append(value)
				print('Child', value, 'was successfully added. Children:', self.data)
			else: print('Child', value, 'could not be added.')

	def findNode(self, data):
		if data in self.data: return self._find_parent(data)
		print(data, 'could not be found.')

	def findLCA(self, bonnie, clyde):
		bonnie, clyde = self.findNode(bonnie), self.findNode(clyde)
		parents_bonnie = self._parents_iterator(bonnie)
		parents_clyde = self._parents_iterator(clyde)
		_lca = self.root.data
		if _lca != None:
			for parent in parents_bonnie:
				if parent in parents_clyde:
					_lca = parent
					break
		return _lca

	def _find_parent(self, data):
		for child in self.nodes:
			if child.data == data:
				return child

	def _parents_iterator(self, node, parents=[]):
		if node.parent == None: return parents
		parents.append(node.parent.data)
		parents = self._parents_iterator(node.parent, parents=parents)
		return parents

if __name__ == '__main__':
	t = Tree()
	t.addNode('George')
	t.addNode('Lucas', parent='George')
	t.addNode('Bonnie', parent='Lucas')
	t.addNode('Georgette', parent='Lucas')
	t.addNode('Clyde', parent='Georgette')
	print(t.findLCA('Bonnie', 'Clyde'))

File: Miscellaneous/DataStructures/Lab2/tree.java
import java.util.ArrayList;

public class tree{
	private Node root;
	private ArrayList<Node> nodes;

	public tree(){
		this.root = null;
		this.nodes = new ArrayList();
	}

	public Node findNode(String data){
		for(Node child: this.nodes)
			if(child.data == data) return child;
		return null;
	}

	public ArrayList<String> listParents(Node child){
		ArrayList<String> parents = new ArrayList();
		Node curr = child;
		while(curr != null){
			parents.add(curr.data);
			curr = curr.parent;
		}
		return parents;
	}

	public void addNode(String data, String parent){
		Node newNode = new Node(data, findNode(parent));
		if(this.root == null) this.root = newNode;
		this.nodes.add(newNode);
	}

	public String findLCA(String bonnie, String clyde){
		ArrayList<String> parents_bonnie = listParents(findNode(bonnie));
		ArrayList<String> parents_clyde = listParents(findNode(clyde));
		for(String s1: parents_bonnie)
			for(String s2: parents_clyde)
				if(s1.equals(s2)) return s1;
		return null;
	}

	public static void main(String[] args) {
		tree t = new tree();
		t.addNode("George", null);
		t.addNode("Lucas", "George");
		t.addNode("Bonnie", "Lucas");
		t.addNode("Georgette", "Lucas");
		t.addNode("Clyde", "Georgette");
		System.out.println(t.findLCA("Bonnie", "Clyde"));
	}
}

class Node{
	Node parent;
	ArrayList<Node> children;
	String data;

	public Node(String value, Node parent){
		this.parent = parent;
		this.children = new ArrayList();
		this.data = value;
	}
}